rpi-eep/src/lib.rs

```diff
@@ -0,0 +1,393 @@
+        }
+        len
+    }
+
+    fn to_bytes(&self, buf: &mut Vec<u8>) {
+        let signature = 0x6950_2d52u32;
+        buf.extend(signature.to_le_bytes());
+        // version
+        buf.push(1);
+        // reserved
+        buf.push(0);
+        // numatoms
+        buf.extend((self.atoms.len() as u16).to_le_bytes());
+        // eeplen
+        buf.extend((self.len() as u32).to_le_bytes());
+        for atom in &self.atoms {
+            atom.to_bytes(buf);
+        }
+    }
+}
+
+impl Default for EEP {
+    fn default() -> Self {
+        EEP::new()
+    }
+}
+
+#[derive(Debug)]
+pub enum EEPAtomData {
+    /// vendor info (0x0001)
+    VendorInfo(EEPAtomVendorData),
+    /// GPIO map (0x0002)
+    GpioMap(gpio_map::EEPAtomGpioMapData),
+    /// Linux device tree blob (0x0003)
+    LinuxDTB(EEPAtomLinuxDTBData),
+    /// manufacturer custom data (0x0004)
+    ManufCustomData(EEPAtomCustomData),
+}
+
+impl ToBytes for EEPAtomData {
+    fn len(&self) -> usize {
+        match self {
+            EEPAtomData::VendorInfo(data) => data.len(),
+            EEPAtomData::GpioMap(data) => data.len(),
+            EEPAtomData::LinuxDTB(data) => data.len(),
+            EEPAtomData::ManufCustomData(data) => data.len(),
+        }
+    }
+    fn to_bytes(&self, buf: &mut Vec<u8>) {
+        match self {
+            EEPAtomData::VendorInfo(data) => data.to_bytes(buf),
+            EEPAtomData::GpioMap(data) => data.to_bytes(buf),
+            EEPAtomData::LinuxDTB(data) => data.to_bytes(buf),
+            EEPAtomData::ManufCustomData(data) => data.to_bytes(buf),
+        };
+    }
+}
+
+/// This enum implements the Atom Types
+///
+/// [Atom Types](https://github.com/raspberrypi/hats/blob/9616b5cd2bdf3e1d2d0330611387d639c1916100/eeprom-format.md#atom-types):
+/// ```text
+/// 0x0000 = invalid
+/// 0x0001 = vendor info
+/// 0x0002 = GPIO map
+/// 0x0003 = Linux device tree blob
+/// 0x0004 = manufacturer custom data
+/// 0x0005-0xfffe = reserved for future use
+/// 0xffff = invalid
+/// ```
+#[derive(Clone, Copy, Debug)]
+#[repr(u8)]
+pub enum EEPAtomType {
+    VendorInfo = 0x0001,
+    GpioMap = 0x0002,
+    LinuxDTB = 0x0003,
+    ManufCustomData = 0x0004,
+}
+
+/// This struct implements the Atom Structure
+///
+/// [Atom Structure](https://github.com/raspberrypi/hats/blob/9616b5cd2bdf3e1d2d0330611387d639c1916100/eeprom-format.md#atom-structure):
+/// ```text
+/// Bytes   Field
+/// 2       type        atom type
+/// 2       count       incrementing atom count
+/// 4       dlen        length in bytes of data+CRC
+/// N       data        N bytes, N = dlen-2
+/// 2       crc16       CRC-16 of entire atom (type, count, dlen, data)
+/// ```
+#[derive(Debug)]
+pub struct EEPAtom {
+    atype: EEPAtomType,
+    /// incrementing atom count
+    count: u16,
+    // length in bytes of data+CRC
+    //dlen: u32,
+    /// N bytes, N = dlen-2
+    data: EEPAtomData,
+    // CRC-16 of entire atom (type, count, dlen, data)
+    //crc16: u16,
```

Comment:
dlen and crc16 are not needed? will they be added in a future update?
If not needed, please add a comment, why.

Reply:
done